File: Server/src/main/java/in/slyther/network/LinkingContext.java
package in.slyther.network;

import in.slyther.gameobjects.GameObject;

import java.util.HashMap;
import java.util.Map;

public class LinkingContext {
    private Map<Integer, GameObject> networkIdToGameObjectMap = new HashMap<>();
    private Map<GameObject, Integer> gameObjectToNetworkIdMap = new HashMap<>();


    public GameObject getGameObject(int networkId) {
        if (!networkIdToGameObjectMap.containsKey(networkId)) {
            return null;
        }
        return networkIdToGameObjectMap.get(networkId);
    }


    public int getNetworkId(GameObject go, boolean shouldCreateIfNotFound) {
        if (gameObjectToNetworkIdMap.containsKey(go)) {
            return gameObjectToNetworkIdMap.get(go);
        } else if (shouldCreateIfNotFound) {
            addGameObject(go);
            return gameObjectToNetworkIdMap.get(go);
        }
        return -1;
    }


    private int nextNetworkId = 1;

    private void addGameObject(GameObject go) {
        addGameObject(go, nextNetworkId++);
    }


    public void addGameObject(GameObject go, int networkId) {
        networkIdToGameObjectMap.put(networkId, go);
        gameObjectToNetworkIdMap.put(go, networkId);
    }


    public void removeGameObject(GameObject go) {
        if (!gameObjectToNetworkIdMap.containsKey(go)) {
            return;
        }

        int networkId = gameObjectToNetworkIdMap.get(go);
        networkIdToGameObjectMap.remove(networkId);
        gameObjectToNetworkIdMap.remove(go);
    }
}

File: Client/Assets/Scripts/Global/Engine/LinkingContext.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public class LinkingContext : MonoBehaviour {
	private Dictionary<int, INetworkGameObject> networkIdToGameObjectDictionary = new Dictionary<int, INetworkGameObject>();
	private Dictionary<INetworkGameObject, int> gameObjectToNetworkIdDictionary = new Dictionary<INetworkGameObject, int>();


	public INetworkGameObject GetGameObject(int networkId)
	{
		if (!networkIdToGameObjectDictionary.ContainsKey(networkId)) {
			return null;
		}
		return networkIdToGameObjectDictionary[networkId];
	}


	public int GetNetworkId(INetworkGameObject go, bool shouldCreateIfNotFound)
	{
		if (gameObjectToNetworkIdDictionary.ContainsKey(go)) {
			return gameObjectToNetworkIdDictionary[go];
		} else if (shouldCreateIfNotFound) {
			AddGameObject(go);
			return gameObjectToNetworkIdDictionary[go];
		}
		return 0;
	}


	private int nextNetworkId = 1;

	private void AddGameObject(INetworkGameObject go)
	{
		AddGameObject(go, nextNetworkId++);
	}


	public void AddGameObject(INetworkGameObject go, int networkId)
	{
		networkIdToGameObjectDictionary.Add(networkId, go);
		gameObjectToNetworkIdDictionary.Add(go, networkId);
	}


	public void RemoveGameObject(INetworkGameObject go)
	{
		if (!gameObjectToNetworkIdDictionary.ContainsKey(go)) {
			return;
		}

		int networkId = gameObjectToNetworkIdDictionary[go];
		networkIdToGameObjectDictionary.Remove(networkId);
		gameObjectToNetworkIdDictionary.Remove(go);
	}
}
